### PR TITLE
feat(layout): add Rect::split method

### PR DESCRIPTION
**Pull request by joshka**
This method splits a Rect and returns a fixed-size array of the
resulting Rects. This allows the caller to use array destructuring
to get the individual Rects.

```rust
use Constraint::*;
let layout = &Layout::vertical([Length(1), Min(0)]);
let [top, main] = area.split(&layout);
```

### CODE REVIEW

**Review by orhun — APPROVED** [review]

**Comment by joshka** [comment]
I want two approvals on this one - mainly for working out whether the naming makes sense.

**Comment by kdheepak** [comment]
In terms of names, I would have preferred `split_into()` if `area` was stored in the layout, i.e.

```rust
let [top, main] =
    Layout::vertical([Constraint::Length(1), Constraint::Min(0)])
    .split_array(area);
```

versus

```rust
let [top, main] =
    Layout::vertical([Constraint::Length(1), Constraint::Min(0)])
    .area(area)
    .split_into();
```

With the functionality as is, I think the name `split_array()` is great!

**Comment by kdheepak** [comment]
I have a small preference for `.split_into_array(area)` as it would be more descriptive and representative of what the function does, i.e.

```rust
layout.split(area)
    .to_vec()
    .try_into()
    .unwrap()
```

is replaced by this:

```rust
layout.split_into_array(area)
```

**Comment by Valentin271** [comment]
I like `split_into_array` as a descriptive name but perhaps it feels too verbose.

> it feels too verbose

Once merged, I assume this will be the main splitting method used.

**Comment by kdheepak** [comment]
Maybe it is worth introducing a breaking change for this and replacing `.split(area)` with this?

**Comment by joshka** [comment]
I'm 👎 on breaking .spilt - 1.1K calls
- https://github.com/search?q=ratatui+layout%3A%3A+.split%28+lang%3Arust+NOT+repo%3Aratatui-org%2Fratatui&type=code

But what about split_area() as a compromise on verbosity?

>Once merged, I assume this will be the main splitting method used.

There's a PR that shows how the code would look using this https://github.com/ratatui-org/ratatui/pull/731

**Comment by kdheepak** [comment]
I like in order of preference:

1. `.split_into(area)`: it's as short as we can get while being descriptive, it hints at what's going on inside the function (i.e. `try_into`)
2. `.split_into_array(area)`: it's descriptive but verbose for something that might be the default way people use this function
3. `.split_array(area)`: it's short but I don't think I'd be able to guess what is going on in the function? It could be misinterpreted as `split the argument which is an array` which is why I like `.split_into_array(area)` more than this
4. `.split_area(area)`: it's short but repeats the argument name in the function name so it falls down the list

**Comment by joshka** [comment]
I think 1,2, and 3 are all problematic because they imply that the area is the (linguistic) object associated with the verb (split_into, split_into_array, split_array). Area is not the object of the sentence its the subject. Only a bare verb, or a verb phrase that repeats the subject communicates this properly.

Put in a sentence we want:
- The area is split by the layout
Not:
- The layout is split into the area
- The layout is split into an array that is the area
- The layout splits an array that is the area

**Comment by joshka** [comment]
Chatting on Discord - @kdheepak made the observation that this then leads to:
```rust
area.split(Layout::horizontal([constraints]));
```
which I've added.
This now also includes the helper methods too.

File: src/layout.rs
Comment (joshka) [inline src/layout.rs:341]: I don't think we can change the new() impl to use Into instead of AsRef, as it would break too much code. Need to investigate the impact. These methods are not yeat released, so we can break code easily.

Making this accpet Into<Constraint> allows for:
```rust
Layout::horizontal([1,2,3,4])
```

File: src/layout.rs
Comment (joshka) [inline src/layout.rs:822]: Do we want both this and min?

**Comment by joshka** [comment]
Needs:
- docs
- updated changelog message
- examples in website to be changed to use this approach

**Review by kdheepak — APPROVED** [review]
This looks good to me! 

But maybe useful to get one more review for feedback on `area.split(&layout)`?

**Comment by joshka** [comment]
> This looks good to me!
> 
> But maybe useful to get one more review for feedback on `area.split(&layout)`?

Yep - definitely want one on this. I'm going to rebase and rewrite the commit for the CHANGELOG also.

**Comment by joshka** [comment]
rebased and commit message / PR text updated.
There's a build failure that is fixed in #743

**Review by Valentin271 — APPROVED** [review]
Would be nice if we could write

```rust
area.split([Constraint::Length(10), Constraint::Length(10)]);
area.split([10, 10]);
```

Not too sure about the second one but the first one is definitely something I'd like.

Approving as this can be part of another PR.

**Review by orhun — APPROVED** [review]

**Comment by joshka** [comment]
> Would be nice if we could write
> 
> ```rust
> area.split([Constraint::Length(10), Constraint::Length(10)]);
> area.split([10, 10]);
> ```
> 
> Not too sure about the second one but the first one is definitely something I'd like.
> 
> Approving as this can be part of another PR.

Direction to split is necessary, so perhaps `Rect::split_horizontal()`:
```rust
fn split_horizontal<const N: usize, T: IntoIterator<Item = Into<Constraint>>(self, constraints: T) -> [Rect; N] {
    self.split(Layout::horizontal(constraints));
}
```
But at that point we're just losing a single word (Layout), which is a good one to keep as it makes it a little more obvious what the [1,2,3] does. I like there to be a solid obvious way to do something. Do you think having both `split(layout)` and `split_horizontal(constraints)` might add one too many ways that would make it difficult to choose which method is the right one to use?

**Comment by Valentin271** [comment]
> it makes it a little more obvious what the [1,2,3] does

I wasn't too sure about that, let's forget that for now.

> Do you think having both split(layout) and split_horizontal(constraints) might add one too many ways that would make it difficult to choose which method is the right one to use?

Probably yes, this is confusing.

---

I though of my suggestion as

```rust
fn split<const N: usize, L: Into<Layout>>(self, layout: L)
```

That way you can still write `.split(layout)` which is the "solid obvious way", and is good if you want to store layout (for configuration perhaps). And for quick and easy split use `Into`.

That would require implementing `From<Vec<Constraint>> for Layout` or similar. But overall it makes sense to convert a bunch of constraint to a layout.

> Direction to split is necessary

As for direction I'd just use the default one. Use layout if you want something else or

**Comment by joshka** [comment]
I'd prefer to avoid defaulting the direction implicitly as it makes the code difficult to read unless you know what that default is. Happy to do it with `default(), but it should be explicit otherwise.

**Comment by Valentin271** [comment]
Alright lets leave it as is for now. On second thought after reading the updated examples it seems fine to me.

**Pull request merged**
